gridcomps/History/MAPL_StationSamplerMod.F90

```diff
@@ -0,0 +1,461 @@
+        if (rank==2) then
+           call ESMF_FieldGet(src_field,farrayptr=p_src_2d,_RC)
+           dst_field = ESMF_FieldCreate(this%esmf_ls,name=xname, &
+                typekind=ESMF_TYPEKIND_R4,_RC)
+           call ESMF_FieldGet(dst_field,farrayptr=p_dst_2d,_RC)
+           call this%regridder%regrid(p_src_2d,p_dst_2d,_RC)
+           if (mapl_am_i_root()) then
+              call this%formatter%put_var(xname,p_dst_2d,&
+                   start=[1,this%obs_written],count=[this%nstation,1],_RC)                         
+           end if
+           call ESMF_FieldDestroy(dst_field,nogarbage=.true.)
+        else if (rank==3) then
+           call ESMF_FieldGet(src_field,farrayptr=p_src_3d,_RC)
+           call ESMF_FieldGet(src_field,ungriddedLBound=lb,ungriddedUBound=ub,_RC)
+           if (this%vdata%lm/=(ub(1)-lb(1)+1)) then
+              lb(1)=1
+              ub(1)=this%vdata%lm
+           end if
+           dst_field = ESMF_FieldCreate(this%esmf_ls,name=xname,&
+                typekind=ESMF_TYPEKIND_R4,ungriddedLBound=lb,ungriddedUBound=ub,_RC)
+           call ESMF_FieldGet(dst_field,farrayptr=p_dst_3d,_RC)
+           !!if (this%vdata%regrid_type==VERTICAL_METHOD_ETA2LEV) then
+           !!   allocate(p_new_lev(size(p_src_3d,1),size(p_src_3d,2),this%vdata%lm),_STAT)
+           !!   call this%vdata%regrid_eta_to_pressure(p_src_3d,p_new_lev,_RC)
+           !!   call this%regridder%regrid(p_new_lev,p_dst_3d,_RC)
+           !!else
+           call this%regridder%regrid(p_src_3d,p_dst_3d,_RC)
+           !!end if
+           !! 
+           if (mapl_am_i_root()) then
+              nx=size(p_dst_3d,1); nz=size(p_dst_3d,2); allocate(arr(nz, nx))
+              arr=reshape(p_dst_3d,[nz,nx],order=[2,1])
+              call this%formatter%put_var(xname,arr,&
+                   start=[1,1,this%obs_written],count=[nz,nx,1],_RC)
+              !note:     lev,station,time
+           end if
+           call ESMF_FieldDestroy(dst_field,nogarbage=.true.)
+        else
+           _FAIL('grid2LS regridder: rank > 3 not implemented')
+        end if  ! rank
+     enddo      ! fieldCount
+     _RETURN(_SUCCESS)
+   end subroutine append_file
+
+   
+   subroutine create_file_handle(this,filename,rc)
+     class(StationSampler), intent(inout) :: this
+     character(len=*), intent(inout) :: filename  ! for ouput nc
+     integer, optional, intent(out) :: rc
+     type(variable) :: v
+     integer :: status
+
+     this%ofile = trim(filename)
+     v = this%time_info%define_time_variable(_RC)
+     call this%fmd%modify_variable('time',v,_RC)
+     !
+     if (mapl_am_I_root()) then
+        call this%formatter%create(trim(filename),_RC)
+        call this%formatter%write(this%fmd,_RC)
+        call this%formatter%put_var('longitude',this%lons,_RC)
+        call this%formatter%put_var('latitude',this%lats,_RC)
+        call this%formatter%put_var('station_id',this%station_id,_RC)                
+     end if
```

Comment:
I'm on a new campaign to reduce long conditionals (esp. ELSE and nested).   You're bad luck that I'm asking you to fix it here.    I also need to create a new macro so that the new conditional and the `_RETURN` can be on the same line.  But that's on me.
```suggestion
     if (.not. mapl_am_I_root()) then
        _RETURN(_SUCCESS)
     end if
     
     call this%formatter%create(trim(filename),_RC)
     call this%formatter%write(this%fmd,_RC)
     call this%formatter%put_var('longitude',this%lons,_RC)
     call this%formatter%put_var('latitude',this%lats,_RC)
     call this%formatter%put_var('station_id',this%station_id,_RC)                
```

Reply:
I think this case here does not need to the twist, nonetheless, I have added in your implementation.

Reply:
I realize that here the i_am_root did not return rc yet, so we still need a ` _RETURN(_SUCCESS)` before end subroutine.